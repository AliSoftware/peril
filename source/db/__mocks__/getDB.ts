import { DatabaseAdaptor, GitHubInstallation } from ".."

// Caches per test file, so you can import it and check from the import
// If this is an issue, don't use this mock :D
let perTestFileMock: MockDB

// A set of type alias' to make the below code readable.
type NullFunc = () => Promise<void>
type InstallationIDToNullInstallation = (installationID: number) => Promise<GitHubInstallation | null>
type InstallationIDToInstallation = (installationID: number) => Promise<GitHubInstallation>
type InstallationToVoid = (installationID: number) => Promise<void>

// Take the DB and rewrite all of its functions to be both the original version and potentially the
// jest mocked version of it.
export interface MockDB extends DatabaseAdaptor {
  setup: NullFunc & jest.Mock<NullFunc>
  getInstallation: InstallationIDToNullInstallation & jest.Mock<InstallationIDToNullInstallation>
  updateInstallation: InstallationIDToNullInstallation & jest.Mock<InstallationIDToNullInstallation>
  saveInstallation: InstallationIDToInstallation & jest.Mock<InstallationIDToNullInstallation>
  deleteInstallation: InstallationToVoid & jest.Mock<InstallationToVoid>
}

// Returns a MockDB which you'll have to alias to alas, because the type system doesn't know
// that jest's mocking system will return a MockDB instead of a typical DatabaseAdaptor

export const getDB = (): MockDB => {
  if (perTestFileMock) {
    return perTestFileMock
  }

  perTestFileMock = {
    getInstallation: jest.fn(),
    deleteInstallation: jest.fn(),
    saveInstallation: jest.fn(),
    updateInstallation: jest.fn(),
    setup: jest.fn(),
  }

  return perTestFileMock
}
