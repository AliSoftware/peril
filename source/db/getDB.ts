// TODO: Is this too fancy and needs fixing?
// e.g. having a get db function
//

import logger from "../logger"
import jsonDB from "./json"
import postgres from "./postgres"

import { DatabaseAdaptor } from "./index"

import { DATABASE_JSON_FILE, DATABASE_URL } from "../globals"

let exposedDB: DatabaseAdaptor = null as any
if (DATABASE_JSON_FILE) {
  logger.info(`Using ${DATABASE_JSON_FILE} as a JSON db`)
  exposedDB = jsonDB(DATABASE_JSON_FILE)
} else if (DATABASE_URL) {
  logger.info(`Using ${DATABASE_URL} as a postgres db`)
  exposedDB = postgres
}

if (exposedDB) {
  exposedDB.setup()
}

export default exposedDB
