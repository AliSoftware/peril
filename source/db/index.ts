import winston from "../logger"
import { GitHubUser } from "./types"

import { GitHubInstallationSettings } from "./GitHubRepoSettings"
import jsonDB from "./json"
import postgres from "./postgres"

/**
 * Should look like one of the following:
 * - "dangerfile.js"
 * - "/path/to/dangerfile.ts"
 * - "repo/slug@to/dangerfile.ts"
 */
export type DangerfileReferenceString = string

/**
 * An event string would be something like:
 * - "pull_request.*"
 * - "pull_request.updated"
 * - "ping"
 * - "issue.*"
 */
export type PerilEventString = string

/** An individual integration of Danger via Peril, this is like the org */
export interface GitHubInstallation {
  /**
   * The ID Integration, this is used when talking to GitHub mainly, but is used
   * as a unique ID in our db
   */
  id: number
  /**
   * In our DB this is represented as a JSON type, so you should anticipate have settings
   * as a nullable type. These are the entire installation settings.
   */
  settings: GitHubInstallationSettings

  /** Having rules in here would mean that it would happen on _any_ event, another JSON type in the DB */
  rules: RunnerRuleset

  /**
   * Scheduled tasks to run using a cron-like syntax.
   *
   * This uses [node-schedule](https://github.com/node-schedule/node-schedule) under the hood. The
   * object is similar to the rules section, in that you define a cron-string with the following format:
   *
   *    *    *    *    *    *    *
   *    ┬    ┬    ┬    ┬    ┬    ┬
   *    │    │    │    │    │    |
   *    │    │    │    │    │    └ day of week (0 - 7) (0 or 7 is Sun)
   *    │    │    │    │    └───── month (1 - 12)
   *    │    │    │    └────────── day of month (1 - 31)
   *    │    │    └─────────────── hour (0 - 23)
   *    │    └──────────────────── minute (0 - 59)
   *    └───────────────────────── second (0 - 59, OPTIONAL)
   *
   * Which would look something like:
   *
   *    "scheduler": {
   *      "0 0 12 * * ?": "schedule/daily_at_twelve.ts",
   *      "0 9 * * 1-5": "schedule/weekday_wakeup_email.ts"
   *    }
   *
   * in practice. There's a lot of great resources on the net showing the general syntax.
   */
  scheduler: RunnerRuleset

  /**
   * A set of repos and their additional event hooks, these are
   * in addition to the ones provided by `"rules"` which are applied
   * to every repo.
   *
   *     "repos" : {
   *       "orta/ORStackView": {
   *          "issue.created": "orta/peril@lock_issues.ts"
   *       }
   *     }
   *
   */
  repos: UniqueRepoRuleset
}

export interface UniqueRepoRuleset {
  [name: string]: RunnerRuleset
}

export interface RunnerRuleset {
  [name: string]: DangerfileReferenceString
}

export interface GithubRepo {
  /** UUID */
  id: number
  /** The installation that this repo is connected to */
  installationID: number
  /** The full name of the slug for this repo, note these could not be unique (GitHub Enterprise?) */
  fullName: string
  /** Runner rules ID reference, another JSON type in the DB */
  rules: RunnerRuleset
}

export interface DatabaseAdaptor {
  /** A once per server start setup function */
  setup: () => Promise<void>

  /** Gets an integrations settings */
  getInstallation: (installationID: number) => Promise<GitHubInstallation | null>
  /** Saves an Integration */
  saveInstallation: (installation: GitHubInstallation) => Promise<any>
  /** Deletes the operation */
  deleteInstallation: (installationID: number) => Promise<void>

  /** Gets an optional repo out of the installation settings */
  getRepo: (installationID: number, repoName: string) => Promise<GithubRepo | null>
  /** Saves a Repo */
  saveGitHubRepo: (repo: GithubRepo) => Promise<any>
  /** Deletes a repo from the  */
  deleteRepo: (installationID: number, repoName: string) => Promise<null | void>
}
