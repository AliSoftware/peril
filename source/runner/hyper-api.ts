import * as aws4 from "hyper-aws4"
import fetch from "node-fetch"
import { HYPER_ACCESS_KEY, HYPER_FUNC_NAME, HYPER_SECRET_KEY } from "../globals"
import logger from "../logger"

// For usage, check out these tests, https://github.com/Tim-Zhang/hyper-aws4/blob/master/test/unit.js

let funcUUID: string | null = null

// https://docs.hyper.sh/Reference/API/2016-04-04%20[Ver.%201.23]/Func/index.html

export const hyper = (path: string, method: "GET" | "PUT" | "POST", body?: any) => {
  const signOption: any = {
    url: "https://us-west-1.hyper.sh/" + path,
    method,
    credential: {
      accessKey: HYPER_ACCESS_KEY,
      secretKey: HYPER_SECRET_KEY,
    },
  }

  if (body) {
    signOption.body = JSON.stringify(body)
  }

  const headers = aws4.sign(signOption)
  const options: any = { method: signOption.method, headers }

  if (body) {
    options.body = JSON.stringify(body)
  }

  return fetch(signOption.url, options).then(res => {
    if (res.ok) {
      return res.json()
    } else {
      return res.text().then(err => {
        logger.error(err)
        throw err
      })
    }
  })
}

// https://docs.hyper.sh/Reference/API/2016-04-04%20[Ver.%201.23]/Func/update.html
export const callHyperFuncUpdate = () => hyper(`func/${HYPER_FUNC_NAME}`, "PUT")

// https://docs.hyper.sh/Reference/API/2016-04-04%20[Ver.%201.23]/Func/call.html
export const callHyperFunction = async (body: any) => {
  // Use the API to grab the UUID needed for the hyper call
  if (!funcUUID) {
    const deets = await getHyperFuncInfo()
    logger.info("d", deets)
    funcUUID = deets.UUID
  }

  return hyper(`func/call/${HYPER_FUNC_NAME}/${funcUUID}`, "POST", body)
}

interface FuncInfo {
  Name: string
  ContainerSize: string
  Timeout: number
  UUID: string
  Created: string
}

// https://docs.hyper.sh/Reference/API/2016-04-04%20%5BVer.%201.23%5D/Func/inspect.html
export const getHyperFuncInfo = () => hyper(`funcs/${HYPER_FUNC_NAME}`, "GET") as Promise<FuncInfo>
