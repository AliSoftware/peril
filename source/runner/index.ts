// @ts-ignore
// tslint:disable-next-line:no-var-requires
const Module = require("module")

const oldRequire = Module.prototype.require

Module.prototype.require = function(id: string) {
  const ts = process.hrtime()
  const res = oldRequire.call(this, id)
  const t = process.hrtime(ts)
  // tslint:disable-next-line:no-console
  console.log("require('%s') took %s ms", id, t[0] * 1000 + t[1] / 1e6)
  return res
}

import * as getSTDIN from "get-stdin"
import nodeCleanup = require("node-cleanup")

try {
  // tslint:disable-next-line:no-var-requires
  const logger = require("../logger").default
  logger.info("Started Runner")

  // We do not trust that the import paths will not provide errors
  // this is my fault for being too smart, and can slowly get fixed in the
  // future. Today however, running Peril's runner in the process
  // with no env vars can be drastically different, and cause unforeseen crashes
  // which don't show up in hyper.
  //
  // So we catch all errors, and thus the upcoming require instead of an import.

  // tslint:disable-next-line:no-var-requires
  const run = require("./run").run
  logger.info("Getting STDIN")

  // Provide a timeout mechanism for the STDIN from the hyper func host
  let foundDSL = false
  getSTDIN().then(stdin => {
    foundDSL = true
    run(stdin)
  })

  // Wait till the end of the process to print out the results. Will
  // only post the results when the process has succeeded, leaving the
  // host process to create a message from the logs.
  nodeCleanup((exitCode, signal) => {
    logger.info(`Process has finished with ${exitCode} ${signal}`)

    // TODO: Failure cases?
    // logger.info("TODO")
    // runtimeEnv.results

    return undefined
  })

  process.on("unhandledRejection", (error: Error) => {
    logger.error("unhandledRejection:", error.message, error.stack)
    process.exitCode = 1
  })

  // Add a timeout so that CI doesn't run forever if something has broken.
  setTimeout(() => {
    if (!foundDSL) {
      logger.error("Timeout: Failed to get the Peril DSL after 2 seconds")
      process.exitCode = 1
      process.exit(1)
    }
  }, 2000)
} catch (error) {
  const err = error as Error
  // tslint:disable-next-line:no-var-requires
  const logger = require("../logger").default
  logger.error(`Error ${err.name} in the runner: ${err.message}\n${err.stack}`)
  throw error
}
