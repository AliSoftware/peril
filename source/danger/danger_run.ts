import { DangerfileReferenceString, RunnerRuleset } from "../db"

/**
 * Think about it, we can't provide the same DSL
 * to a PR as we send to an Issue, or to User creation, the lack of
 * reference to a PR means that we can't do work like finding diffs.
 */
export enum dsl {
  /** What, for years, has been the "Danger DSL" */
  pr,
  /** Take whatever JSON triggered this run and use that as the `github.` DSL */
  import,
}

/** Can Danger reply inline? */
export enum feedback {
  /** Is there a way in which Danger can provide any feedback? */
  commentable,
  /** Can only execute the JS, no feedback into an issue as the event doesn't correlate to one */
  silent,
}

/** Represents runs that Danger should do based on Rules and Events */
export interface DangerRun extends RepresentationForURL {
  /** What event name triggered this */
  event: string
  /** What action inside that event trigger this run */
  action: string | null
  /** What type of DSL should the run use? */
  dslType: dsl
  /** Can Danger provide commentable feedback? */
  feedback: feedback
}

/** Takes an event and action, and defines whether to do a dangerfile run with it. */
export const dangerRunForRules = (
  event: string,
  action: string | null,
  rule: RunnerRuleset | undefined | null
): DangerRun | null => {
  // tslint:disable-line
  // Can't do anything with nothing
  if (!rule) {
    return null
  }

  // We can just see if anything exists at the right places,
  // and return the first right object
  const isDirect = rule[event]
  const globsAll = rule[event + ".*"]
  const eventDotAction = action && rule[event + "." + action]
  const path = isDirect || globsAll || eventDotAction

  // Bail, we can't do anything
  if (!path) {
    return null
  }

  return {
    action,
    dslType: dslTypeForEvent(event),
    event,
    ...dangerRepresentationforPath(path),
    feedback: feedbackTypeForEvent(event),
  }
}

interface RepresentationForURL {
  dangerfilePath: string
  branch: string
  repoSlug: string | undefined
}

/** Takes a DangerfileReferenceString and lets you know where to find it globally */
export const dangerRepresentationforPath = (value: DangerfileReferenceString): RepresentationForURL => {
  return {
    branch: value.includes("#") ? value.split("#")[0] : "master",
    dangerfilePath: value.split("#")[0],
    repoSlug: value.includes("@") ? value.split("@")[0] : undefined,
  }
}

/** What type of DSL should get used for the Dangerfile eval? */
export const dslTypeForEvent = (event: string): dsl => {
  if (event === "pull_request") {
    return dsl.pr
  }
  return dsl.import
}

/** What events can we provide feedback inline with? */
// Build system mentions?
export const feedbackTypeForEvent = (event: string): feedback => {
  if (event === "pull_request" || event === "issues" || event === "issue") {
    return feedback.commentable
  }
  return feedback.silent
}
