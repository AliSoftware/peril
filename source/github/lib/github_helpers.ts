import fetch from "../../api/fetch"
import { GitHubUser } from "../../db/types"
import winston from "../../logger"

export async function isUserInOrg(token: string, user: GitHubUser, org: string) {
    // https://developer.github.com/v3/orgs/members/#check-membership
    const res = await api(token, `orgs/${org}/members/${user}`)
    return res.status === 204
}

/**
 * There's definitely a time when you want access to a GitHub file
 * but won't have an auth token to do it yet, this function should
 * help out there, you can provide any auth token you want.
 * Returns either the contents or nothing.
 */
export async function getGitHubFileContents(token: string | null, repoSlug: string, path: string, ref: string | null) {
    const refString = ref ? `ref=${ref}` : ""
    const res = await api(token, `repos/${repoSlug}/contents/${path}?${refString}`)
    const data = await res.json()
    if (res.ok) {
        const buffer = new Buffer(data.content, "base64")
        return buffer.toString()
    } else {
        winston.error("res: " + res.url)
        winston.error("Getting GitHub file failed: " + JSON.stringify(data))
        return ""
    }
}

/**
 * A quick GitHub API client
 */
async function api(token: string | null, path: string, headers: any = {}, body: any = {}, method: string = "GET") {
    if (token) { headers.Authorization = `token ${token}` }

    const baseUrl = process.env.DANGER_GITHUB_API_BASE_URL || "https://api.github.com"
    return fetch(`${baseUrl}/${path}`, {
        body,
        headers: {
            "Accept": "application/vnd.github.machine-man-preview+json",
            "Content-Type": "application/json",
            ...headers,
        },
        method,
    })
}
