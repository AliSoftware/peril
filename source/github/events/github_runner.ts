import * as express from "express"
import winston from "../../logger"

import { PERIL_BOT_USER_ID } from "../../globals"

import { GitHub } from "danger/distribution/platforms/GitHub"
import { GitHubAPI } from "danger/distribution/platforms/github/GitHubAPI"

import { DangerResults } from "danger/distribution/dsl/DangerResults"
import { getTemporaryAccessTokenForInstallation } from "../../api/github"
import { DangerRun, dangerRunForRules, dsl, feedback } from "../../danger/danger_run"
import { executorForInstallation, runDangerAgainstInstallation } from "../../danger/danger_runner"
import db, { GitHubInstallation, GithubRepo } from "../../db"
import { getGitHubFileContents, isUserInOrg } from "../lib/github_helpers"

/**
 * So, these function have a bunch of responsibilities.
 *
 *  - Validating there is an installation ref in the db
 *  - Generating runs for an installation, could be up to two (org + repo) per integration event
 *  - Going from a run to executing Danger for that run
 *  - Handling the varients in a Danger run
 *
 *    - Event is org based (no repo, DSL is event JSON)
 *    - Event is repo based (has a reference to a repo, but nothing to comment on)
 *    - Event is PR based (has a repo + issue, can comment, gets normal DangerDSL)
 *    - Event is issue based (has a repo + issue, can comment, gets event DSL )
 *
 *  - Passing back the feedback results, if we can
 *
 * As you can imagine, this does indeed make it ripe for a good refactoring in the future.
 */

/** Logs */
const log = (message: string) => {
  winston.info(`[runner] - ${message}`)
}

export interface GitHubRunSettings {
  commentableID: number | null
  isRepoEvent: boolean | null
  isTriggeredByUser: boolean
  repo: GithubRepo | null
  repoName: string | null
  triggeredByUsername: string | null
  hasRelatedCommentable: boolean
}

export const setupForRequest = async (req: express.Request): Promise<GitHubRunSettings> => {
  const isRepoEvent = !!req.body.repository
  const repoName = isRepoEvent && req.body.repository.full_name
  const installationID = req.body.installation.id as number
  const isTriggeredByUser = !!req.body.sender
  const hasRelatedCommentable = getIssueNumber(req.body) !== null

  return {
    commentableID: hasRelatedCommentable ? getIssueNumber(req.body) : null,
    hasRelatedCommentable,
    isRepoEvent,
    isTriggeredByUser,
    repo: isRepoEvent ? await db.getRepo(installationID, repoName) : null,
    repoName,
    triggeredByUsername: isTriggeredByUser ? req.body.sender.login : null,
  }
}

export const githubDangerRunner = async (event: string, req: express.Request, res: express.Response, next: any) => {
  const action = req.body.action as string | null
  const installationID = req.body.installation.id as number

  const installation = await db.getInstallation(installationID)
  if (!installation) {
    res.status(404).send(`Could not find installation with id: ${installationID}`)
    return
  }

  const settings = await setupForRequest(req)

  // TODO: Why did I make this call before?
  if (!settings.isRepoEvent) {
    res.status(404).send(`WIP - not built out support for non-repo related events - sorry`)
    return
  }

  const runs = runsForEvent(event, action, installation, settings)
  await runEverything(runs, settings, installation, req, res, next)
}

export function runsForEvent(event: string, action: string | null, installation: GitHubInstallation, settings: any) {
  const installationRun = dangerRunForRules(event, action, installation.rules)
  const repoRun = dangerRunForRules(event, action, settings.repo && settings.repo.rules)
  return [installationRun, repoRun].filter(r => !!r) as DangerRun[]
}

export const runEverything = async (
  runs: DangerRun[],
  settings: any,
  installation,
  req: express.Request,
  res: express.Response,
  next: any
) => {
  // We got no runs ( so there were no rules that correspond to the event)
  if (runs.length === 0) {
    res.status(204).send(`No work to do.`)
    next()
    return
  }

  log(`Event Settings: ${JSON.stringify(settings, null, " ")}`)
  const token = await getTemporaryAccessTokenForInstallation(installation)

  const allResults = [] as DangerResults[]
  for (const run of runs) {
    log(`Running: ${JSON.stringify(run, null, " ")}`)

    const isPR = run.dslType === dsl.pr
    const supportGithubCommentAPIs = run.feedback === feedback.commentable

    log(`Use fullDSL: ${isPR}`)
    log(`supportGithubCommentAPIs: ${supportGithubCommentAPIs}`)

    // Do we need an authenticated Danger GitHubAPI instance so we
    // can leave feedback on an issue?
    let githubAPI = null as GitHubAPI | null
    if (supportGithubCommentAPIs && settings.commentableID && settings.repo) {
      githubAPI = githubAPIForCommentable(token, settings.repo.fullName, settings.commentableID)
      log("Got a GitHub API")
    }

    // Are we being extra paranoid about running Dangerfiles?
    // Ideally this can move to detect if the PR changed the Dangerfile also
    const onlyOrgPR = installation.settings.onlyForOrgMembers && isPR
    if (onlyOrgPR && githubAPI && settings.repoName && settings.triggeredByUsername) {
      log("Checking if user is in org")
      const org = settings.repoName.split("/")[0]
      const userInOrg = await isUserInOrg(token, settings.triggeredByUsername, org)
      if (!userInOrg) {
        res.status(403).send(`Not running because ${settings.triggeredByUsername} is not in ${org}.`)
        return
      }
    }

    // In theory only a PR requires a custom branch, so we can check directly for that
    // in the event JSON and if it's not there then use master
    // prioritise the run metadata
    const repoForDangerfile = run.repoSlug || settings.repoName
    const dangerfileBranchForPR = isPR ? req.body.pull_request.head.ref : null
    const neededDangerfileIsSameRepo = isPR ? run.repoSlug === req.body.pull_request.head.repo.full_name : false
    const branch = neededDangerfileIsSameRepo ? dangerfileBranchForPR : null

    const file = await getGitHubFileContents(token, repoForDangerfile, run.dangerfilePath, branch)
    if (file !== "") {
      const results = await runDangerAgainstInstallation(file, run.dangerfilePath, githubAPI, run.dslType)
      allResults.push(results)
    } else {
      log("Got no github file contents, commenting.")
      const results = {
        repoForDangerfile,
        dangerfileBranchForPR,
        neededDangerfileIsSameRepo,
        onlyOrgPR,
        supportGithubCommentAPIs,
        isPR,
        branch,
      }
      const actualBranch = branch ? branch : "master"
      const message = `Could not find Dangerfile at ${run.dangerfilePath} on ${repoForDangerfile} on branch ${actualBranch}.  
      
    ${JSON.stringify(results)}
      
      `
      allResults.push({ fails: [{ message }], markdowns: [], warnings: [], messages: [] })
    }
  }

  const commentableRun = runs.find(r => r.feedback === feedback.commentable)
  if (commentableRun && allResults.length) {
    const finalResults = mergeResults(allResults)
    log(`Commenting, with results: ${mdResults(finalResults)}`)
    commentOnResults(finalResults, token, settings)
  }

  res.status(200).send(`Run ${runs.length} Dangerfiles`)
}

export const mdResults = (results: DangerResults): string => {
  return `
mds: ${results.markdowns.length}
mds: ${results.messages.length}
warns: ${results.warnings.length}
fails: ${results.fails.length}
  `
}

export const mergeResults = (results: DangerResults[]): DangerResults => {
  return results.reduce(
    (curr: DangerResults, newResults: DangerResults) => {
      return {
        fails: [...curr.fails, ...newResults.fails],
        markdowns: [...curr.markdowns, ...newResults.markdowns],
        messages: [...curr.messages, ...newResults.messages],
        warnings: [...curr.warnings, ...newResults.warnings],
      }
    },
    { fails: [], markdowns: [], warnings: [], messages: [] }
  )
}

export const commentOnResults = async (results: DangerResults, token, settings) => {
  const githubAPI = githubAPIForCommentable(token, settings.repoName, settings.commentableID)
  const exec = executorForInstallation(new GitHub(githubAPI))
  await exec.handleResults(results)
}

// This doesn't feel great, but is OK for now
const getIssueNumber = (json: any): number | null => {
  if (json.pull_request) {
    return json.pull_request.number
  }
  if (json.issue) {
    return json.issue.number
  }
  return null
}

// This doesn't feel great, but is OK for now
const getRepoSlug = (json: any): string | null => {
  if (json.repository) {
    return json.repository.full_name
  }
  return null
}

const githubAPIForCommentable = (token: string, repoSlug: string, issueNumber: number | null) => {
  const githubAPI = new GitHubAPI({ repoSlug, pullRequestID: String(issueNumber) }, token)
  githubAPI.additionalHeaders = { Accept: "application/vnd.github.machine-man-preview+json" }

  // How can I get this from an API, if we cannot use /me ?
  // https://api.github.com/repos/PerilTest/PerilPRTester/issues/5/comments
  // Talked to GH - they know it's an issue.
  githubAPI.getUserID = () => Promise.resolve(parseInt(PERIL_BOT_USER_ID, 10))
  return githubAPI
}

export default githubDangerRunner
