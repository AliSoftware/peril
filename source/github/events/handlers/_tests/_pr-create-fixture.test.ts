jest.mock("../../../../db/getDB")
import { MockDB } from "../../../../db/__mocks__/getDB"
import { getDB } from "../../../../db/getDB"
const mockDB = getDB() as MockDB

jest.mock("../../../../api/github", () => ({
  getTemporaryAccessTokenForInstallation: () => Promise.resolve("token"),
}))

jest.mock("../../../../github/lib/github_helpers", () => ({
  getGitHubFileContents: jest.fn(),
  canUserWriteToRepo: () => true,
}))
import { getGitHubFileContents } from "../../../lib/github_helpers"
const mockGetGitHubFileContents: any = getGitHubFileContents

import { readFileSync, writeFileSync } from "fs"
import { resolve } from "path"
import { dangerRunForRules } from "../../../../danger/danger_run"
import { callHyperFunction } from "../../../../runner/hyper-api"
import { PerilRunnerBootstrapJSON } from "../../../../runner/triggerSandboxRun"
import { setupForRequest } from "../../github_runner"

import { runPRRun } from "../pr"

jest.mock("../../../../runner/hyper-api", () => ({
  callHyperFunction: jest.fn(() => Promise.resolve('{ "CallId": 123 }')),
}))

import { FakePlatform } from "danger/distribution/platforms/FakePlatform"

const mockPlatform = new FakePlatform()
jest.mock("../../../../danger/peril_platform", () => ({
  getPerilPlatformForDSL: () => mockPlatform,
}))

const apiFixtures = resolve(__dirname, "../../_tests/fixtures")
const fixture = (file: string) => JSON.parse(readFileSync(resolve(apiFixtures, file), "utf8"))

it("passes the right args to the hyper functions when it's a PR", async () => {
  mockDB.getInstallation.mockReturnValue({ iID: "123", repos: {}, envVars: { hello: "world" } })

  const body = fixture("pull_request_opened.json")
  const req = { body, headers: { "X-GitHub-Delivery": "123" } } as any
  const settings = await setupForRequest(req, {})

  const dangerfileForRun = "warn(danger.github.api)"
  mockGetGitHubFileContents.mockImplementationOnce(() => Promise.resolve(dangerfileForRun))

  const runSettings = { pull_request: "org/repo@warn_with_api.ts#branch" }
  const run = dangerRunForRules("pull_request", "created", runSettings, body)

  await runPRRun(run, settings, "token", body.pull_request)

  // Take the payload, remove the JWT and save a copy of the JSON into a fixture dir, then snapshot it
  const payload = (callHyperFunction as any).mock.calls[0][0] as PerilRunnerBootstrapJSON
  payload.perilSettings.perilJWT = "[skipped]"
  payload.perilSettings.perilRunID = "[run-id]"
  writeFileSync(__dirname + "/fixtures/PerilRunnerPRBootStrapExample.json", JSON.stringify(payload, null, "  "), "utf8")

  expect(payload).toMatchSnapshot()
})
