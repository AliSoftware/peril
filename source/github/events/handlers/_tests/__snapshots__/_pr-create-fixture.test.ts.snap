// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes the right args to the hyper functions when it's a PR 1`] = `
Object {
  "dslType": "pr",
  "installation": Object {
    "iID": 4766,
    "settings": Object {},
  },
  "paths": Array [
    "org/repo@warn_with_api.ts#branch",
  ],
  "payload": Object {
    "dsl": Object {
      "git": Object {
        "JSONDiffForFile": [Function],
        "JSONPatchForFile": [Function],
        "commits": Array [
          Object {
            "author": Object {
              "date": "1",
              "email": "1",
              "name": "1",
            },
            "committer": Object {
              "date": "1",
              "email": "1",
              "name": "1",
            },
            "message": "456",
            "sha": "123",
            "tree": Object {
              "sha": "123",
              "url": "123",
            },
            "url": "123",
          },
        ],
        "created_files": Array [],
        "deleted_files": Array [],
        "diffForFile": [Function],
        "modified_files": Array [],
        "structuredDiffForFile": [Function],
      },
      "github": Object {},
      "settings": Object {
        "cliArgs": Object {},
        "github": Object {
          "accessToken": "token",
          "additionalHeaders": Object {
            "Accept": "application/vnd.github.machine-man-preview+json",
          },
          "baseURL": undefined,
        },
      },
    },
    "webhook": null,
  },
  "perilSettings": Object {
    "envVars": Object {
      "hello": "world",
    },
    "perilAPIRoot": undefined,
    "perilJWT": "[skipped]",
    "perilRunID": "[run-id]",
  },
}
`;
